[Merged by Bors] - Contextual logging

lrettig opened the pull request:
## Motivation
Closes #2322 

## Changes
- Threads context through network and message handling for debug purposes
- Adds app context-related helper methods
- Adds ability to add context to logging

## Test Plan
- [x] Add unit tests for log context helpers
- [x] Add an integration test that tracks a requestId from top to bottom

## TODO
<!-- This section should be removed when all items are complete -->
- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [ ] Update documentation as needed

lrettig committed:
Add basic support for contextual logging (WIP)

noamnelke commented on p2p/net/network.go:
```diff
@@ -290,27 +293,31 @@ func (n *Net) accept(listen net.Listener) {
 				return
 			}
-			go c.beginEventProcessing()
+			go c.beginEventProcessing(context.TODO())
```
Why not use `context.Background()`? Do you think we should pass a context from the outside here in the future?

lrettig replied:
you're right, this one is not a TODO. fixed.

noamnelke commented on p2p/net/udp.go:
```diff
@@ -259,12 +266,13 @@ func (n *UDPNet) listenToUDPNetworkMessages(listener net.PacketConn) {
 			n.publishNewRemoteConnectionEvent(mconn, node.NewNode(pk, net.ParseIP(host), 0, uint16(iport)))
 			n.addConn(addr, conn)
-			go mconn.beginEventProcessing()
+			go mconn.beginEventProcessing(context.TODO())
```
Same here, it's fine to leave `TODO` if you think we should create an app-level context and pass it in here, just wanted to know your thoughts. In general, I think we should use/pass contexts more broadly and we should probably do it as part of the graceful shutdown work (using Tomb).

lrettig replied:
cleaned this one up too. agree re: graceful shutdown work, context objects can be used for that too, and this work should make that easier, but for now I'm focused just on logging.

noamnelke replied:
>cleaned this one up

?

lrettig replied:
cleaned up as in it's no longer a TODO

noamnelke replied:
yes, it is 😬

noamnelke commented:
Just noticed that it's still a draft... 🤭

lrettig committed:
Clean up context helpers and thread through hare

Finish adding basic infrastructure for contextual logging
Continue threading context down through hare message handling (still
WIP)

lrettig committed:
Make linter happy

lrettig committed:
Fix import cycle, make linter happy again

lrettig committed:
Some minor debugging

lrettig committed:
Fix network listen context

lrettig committed:
Improve listener context generation

lrettig committed:
Adding context to hare stack (WIP)

lrettig committed:
Finish initial rollout of contextual logging

Finished threading hare
Added to top-level and passed into most services

lrettig committed:
Make linter happy again

lrettig committed:
Fix a broken test, minor test cleanup

lrettig committed:
Minor cleanup

lrettig committed:
More minor cleanup

lrettig committed:
Wire up p2p stack with context

And a little more minor cleanup

lrettig committed:
Make linter happy

lrettig committed:
Fix test

lrettig committed:
Merge branch 'develop' into contextual-logging

lrettig committed:
Add context, improve debugging of syncer

lrettig committed:
Context, logging for block listener, syncer

lrettig committed:
Minor logging cleanup and improvements

lrettig committed:
Fix merge conflicts, add context to new block code

lrettig committed:
Improve hare logging and context

Properly reconstruct reqID from hare message from broker

lrettig committed:
Improve hare context and logging

lrettig committed:
Add test for contextual logging

lrettig committed:
Add roundtrip test for hare msg reqID

lrettig committed:
Merge branch 'develop' into contextual-logging

lrettig committed:
Make linter happy again

lrettig committed:
Fix tests

lrettig commented:
bors try

lrettig committed:
Add context to hare outgoing messages

lrettig committed:
Linter

lrettig committed:
More lint and test fixes

lrettig committed:
Fix cherry-pick conflicts

lrettig committed:
Pass context down to msg send on net

lrettig committed:
Improve net debug

lrettig committed:
Standardize formatting of hare msg type string

lrettig committed:
Fix test to match log change

lrettig committed:
Properly handle nil msg

lrettig committed:
Merge branch 'develop' into contextual-logging

lrettig committed:
Fix hare msg_type output

Simplify this change. Rename p2p msg_type field so as not to break
kibana type detection.

Partially revert "Properly handle nil msg"

This reverts commit 5ce329a19fe135ad3053d2763117741db31e7670.

lrettig committed:
Improve context and logging for outgoing messages

noamnelke commented on activation/atxdb.go:
```diff
@@ -746,38 +747,38 @@ func (db *DB) ValidateSignedAtx(pubKey signing.PublicKey, signedAtx *types.Activ
 	}
 
-	if err := syncer.GetPoetProof(atx.GetPoetProofRef()); err != nil {
+	if err := syncer.GetPoetProof(context.TODO(), atx.GetPoetProofRef()); err != nil {
```
```suggestion
	if err := syncer.GetPoetProof(ctx, atx.GetPoetProofRef()); err != nil {
```

Or am I missing something?

If I am, I guess it should still not be `TODO()`

lrettig replied:
already caught and fixed this

noamnelke commented on activation/poetlistener.go:
```diff
@@ -74,7 +75,7 @@ func (l *PoetListener) handlePoetProofMessage(gossipMessage service.GossipMessag
 	}
 
-	gossipMessage.ReportValidation(PoetProofProtocol)
+	gossipMessage.ReportValidation(context.TODO(), PoetProofProtocol)
```
```suggestion
	gossipMessage.ReportValidation(ctx, PoetProofProtocol)
```

lrettig committed:
Improve context and logging for atx and block flow

Remove explicit context and reqID from ReportValidation, and read from
message context instead

Resolves merge conflicts from testnet-126 branch changes

lrettig committed:
Logging

lrettig committed:
Minor cleanup of atx logging

lrettig committed:
Bugfixes

lrettig committed:
Add log context to fetch request

lrettig committed:
Whitespace

lrettig committed:
Improve context and logging in sync stack

lrettig committed:
Bugfix in atxFields

lrettig committed:
Fix interfaces for IsSynced

lrettig committed:
Add context to request handlers

lrettig committed:
Update interfaces

lrettig committed:
Improve logging for fetch queue and workers

lrettig committed:
WIP: timeout on handle fetch failure (with ctx)

lrettig committed:
Improve context and logging in fetch queue stack

lrettig committed:
Bugfixes in fetch queue timeout

AmitShaul commented on activation/activation.go:
```diff
@@ -244,7 +245,7 @@ func (b *Builder) buildNipstChallenge(currentLayer types.LayerID) error {
 // StartPost initiates post commitment generation process. It returns an error if a process is already in progress or
 // if a post has been already initialized
-func (b *Builder) StartPost(rewardAddress types.Address, dataDir string, space uint64) error {
+func (b *Builder) StartPost(ctx context.Context, rewardAddress types.Address, dataDir string, space uint64) error {
```
is Context necessary here?
I don't see any use of it in this function

lrettig replied:
fixed!

AmitShaul commented on activation/poetlistener.go:
```diff
@@ -26,7 +27,7 @@ type PoetListener struct {
 
 // Start starts listening to PoET gossip messages.
-func (l *PoetListener) Start() {
+func (l *PoetListener) Start(ctx context.Context) {
```
same here, is it necessary to have Context here without use?

lrettig replied:
fixed

AmitShaul commented on hare/broker.go:
```diff
@@ -141,77 +142,70 @@ func (b *Broker) validate(m *Message) error {
 				continue
 			}
 
+			// create an inner context object to handle this message
+			messageCtx := ctx
+
+			// try to read stored context
+			if msg.RequestID() == "" {
+				logger.With().Warning("broker received hare message with no registered requestID")
+			} else {
+				messageCtx = log.WithRequestID(ctx, msg.RequestID())
+			}
+
```
I think that moving line 156 into the first if statement on line 159 will be more clear, it's a matter of semantics so it's not an issue if you decide to leave it this way.

lrettig replied:
That won't work, it needs to be in scope outside the if statement

AmitShaul commented on activation/activation.go:
```diff
@@ -3,6 +3,7 @@
 package activation
 
 import (
+	"context"
```
you sometimes import "golang.org/x/net/context" (like in gateway_service.go) and sometimes "context"

lrettig replied:
Thanks for noticing. Actually, I didn't add the context imports to the GRPC code in this PR. It was there from before. The google GRPC stack may be doing something different with context so I think this can be left alone.

noamnelke replied:
It looks like `golang.org/x/net/context` is (unofficially) deprecated. We should probably use the package from the standard library (assuming we require Go 1.7+, otherwise we can't use it at all).

https://github.com/grpc-ecosystem/go-grpc-middleware/issues/175

(doesn't have to be in this PR).

lrettig replied:
okay, will fix this, thanks guys

AmitShaul commented on hare/proposaltracker.go:
```diff
@@ -72,16 +73,19 @@ func (pt *proposalTracker) OnLateProposal(msg *Msg) {
 	// lower ranked proposal on late proposal is a conflict
 	if bytes.Compare(msg.InnerMsg.RoleProof, pt.proposal.InnerMsg.RoleProof) < 0 {
-		pt.With().Info("late lower rank detected", log.String("id_malicious", msg.PubKey.String()))
+		pt.With().Info("late lower rank detected",
+			log.String("id_malicious", msg.PubKey.String()))
```
why the new line?:)

lrettig replied:
:shrug:

AmitShaul commented on sync/validation_queue.go:
```diff
@@ -253,14 +254,12 @@ func (vq *blockQueue) addDependencies(jobID interface{}, blks []types.BlockID, f
-		log.Int("count", len(dependencies)),
-		log.String("job_id", fmt.Sprintf("%v", jobID)))
+	logger.With().Debug("finished adding dependencies",
+		log.Int("count", len(dependencies)))
```
did you intentionally remove the job_id?

lrettig replied:
it's added to the context on line 214, above, so it will appear in all log lines - the beauty of contextual logging :)

AmitShaul commented on p2p/gossip/protocol.go:
```diff
@@ -105,32 +107,31 @@ func (p *Protocol) markMessageAsOld(h types.Hash12) bool {
+func (p *Protocol) processMessage(ctx context.Context, sender p2pcrypto.PublicKey, protocol string, msg service.Data) error {
 	h := types.CalcMessageHash12(msg.Bytes(), protocol)
-	fields := []log.LoggableField{
+	logger := p.WithContext(ctx).WithFields(
 		log.String("from", sender.String()),
 		log.String("protocol", protocol),
```
indentation

lrettig replied:
what do you mean? looks okay to me.

lrettig committed:
Merge branch 'develop' into contextual-logging

lrettig committed:
Tweaks per PR review from Amit

noamnelke commented on cmd/node/node.go:
```diff
@@ -456,7 +449,8 @@ func (app *SpacemeshApp) SetLogLevel(name, loglevel string) error {
 }
 
-func (app *SpacemeshApp) initServices(nodeID types.NodeID,
+func (app *SpacemeshApp) initServices(ctx context.Context,
```
while we pass the `ctx` along, we don't use it for logs in this method - shouldn't we?

(i.e. `logger := app.log.WithContext(ctx)` + `s/app.log/logger/`)

noamnelke commented on cmd/node/node.go:
```diff
@@ -706,37 +702,33 @@ func (app *SpacemeshApp) HareFactory(mdb *mesh.DB, swarm service.Service, sgn ha
-	//app.blockListener.Start()
-	go app.startSyncer()
+func (app *SpacemeshApp) startServices(ctx context.Context) {
+	logger := log.AppLog.WithContext(ctx)
```
I think this is the first use of `log.AppLog` outside of testing code. I don't see how this is any worse than using bare calls logging methods with no logger, but just making sure you're aware. Consider passing a "real" logger in here?

lrettig replied:
done, and done throughout this module - but I think there are probably still a few other scattered places where I did the same thing. We're not very consistent with how log objects are passed around today, that still needs to be improved.

noamnelke commented on layerfetcher/layers.go:
```diff
@@ -172,23 +149,23 @@ func (l *Logic) PollLayer(layer types.LayerID) chan LayerPromiseResult {
+			l.receiveLayerHash(ctx, layer, p, len(peers), nil, err)
 		}
-		err := l.net.SendRequest(LayerBlocksDB, layer.Bytes(), p, receiveForPeerFunc, timeoutFunc)
+		err := l.net.SendRequest(context.TODO(), LayerBlocksDB, layer.Bytes(), p, receiveForPeerFunc, timeoutFunc)
```
missed a spot 🧽

```suggestion
		err := l.net.SendRequest(ctx, LayerBlocksDB, layer.Bytes(), p, receiveForPeerFunc, timeoutFunc)
```

noamnelke commented on layerfetcher/layers.go:
```diff
@@ -243,15 +220,15 @@ func (l *Logic) receiveLayerHash(id types.LayerID, p p2ppeers.Peer, peers int, d
+			l.receiveBlockHashes(ctx, id, nil, err)
 		}
-		err := l.net.SendRequest(LayerHashDB, hash.Bytes(), peer[remainingPeers], receiveForPeerFunc, errFunc)
+		err := l.net.SendRequest(context.TODO(), LayerHashDB, hash.Bytes(), peer[remainingPeers], receiveForPeerFunc, errFunc)
```
Either you consistently miss this spot, or I'm missing something?

```suggestion
		err := l.net.SendRequest(ctx, LayerHashDB, hash.Bytes(), peer[remainingPeers], receiveForPeerFunc, errFunc)
```

noamnelke commented on layerfetcher/layers.go:
```diff
@@ -331,11 +298,11 @@ func (l *Logic) getPoetResult(hash types.Hash32, data []byte) error {
 // blockReceiveFunc handles blocks received via fetch
 func (l *Logic) blockReceiveFunc(data []byte) error {
-	return l.blockHandler.HandleBlockData(data, l)
+	return l.blockHandler.HandleBlockData(context.TODO(), data, l)
```
I guess this should be fixed

noamnelke commented on p2p/message.go:
```diff
@@ -33,6 +33,7 @@ type gossipProtocolMessage struct {
 	sender         p2pcrypto.PublicKey
 	data           service.Data
 	validationChan chan service.MessageValidation
+	requestID      string
```
If I'm understanding the flow correctly, this means we'll be able to log the request ID of a message that originated in the same node as it passes through the processing pipeline before being gossipped (intra-node).

We'll only be able to track messages inter-node for Hare messages, where you've added the request ID to the message on the wire.

Is this the intended behaviour?

lrettig replied:
RequestID isn't added to the message on the wire. Only the `data` member here is serialized.

noamnelke commented on hare/algorithm.go:
```diff
@@ -95,7 +96,8 @@ type StateQuerier interface {
 	*Message
-	PubKey *signing.PublicKey
+	PubKey    *signing.PublicKey
+	RequestID string
```
I'm wondering if the addition of a field to every gossip message on the wire for the sole purpose of easier to follow logs is worth it... 🤔

My intuition is that it isn't worth it and that we should do the best we can with the minimal messages required to operate the protocol.

Have we (the team) discussed this before implementation? Maybe I just don't remember, or was absent, and we did. It seems like a pretty big decision to me, worth discussing.

lrettig replied:
No, that wasn't my intention, and no, we didn't discuss this. RequestIDs (and SessionIDs) are strictly local and have no meaning on the wire, or to another node, so they should not be sent on the wire. The field I added here, and to the gossip message, is just used to convey context across different goroutines/contexts locally. Let me rethink this - I might be able to achieve the same thing a different way, or else, I can wrap the on-the-wire messages in a struct that's passed around locally if necessary. Thanks for catching this.

lrettig replied:
Took a closer look. As with the above comment (gossipProtocolMessage), RequestID is not sent on the wire. Only the inner `Message` member here is serialized.

noamnelke replied:
You're right, I missed this.

noamnelke commented on p2p/net/udp.go:
```diff
@@ -84,7 +84,7 @@ func (n *UDPNet) SubscribeClosingConnections(f func(connection ConnectionWithErr
 	n.clsMutex.RLock()
 	for _, f := range n.closingConnections {
-		f(connection)
+		f(context.TODO(), connection)
```
```suggestion
		f(ctx, connection)
```

lrettig replied:
Won't work, there is no ctx here. I didn't try to add context _everywhere_, including the full p2p stack. That's what context.TODO is for :) Feel free to finish threading it through this stack... (this can be done incrementally in future - e.g., when debugging issues, where it's very helpful)

noamnelke commented on sync/block_listener.go:
```diff
@@ -36,9 +37,9 @@ func (bl *BlockListener) Close() {
 
 // Start starts the main listening goroutine
-func (bl *BlockListener) Start() {
+func (bl *BlockListener) Start(ctx context.Context) {
```
Why do we keep this commented-out code in the repo? (this entire file)

lrettig replied:
No idea, that was part of #2128 - @antonlerner ? I'll remove them here, this is sort of already a "general cleanup" project anyway

noamnelke reviewed (commented):
My main issue with this change is that it changes the Hare gossip messages on the wire for the purpose of nicer logging. I _think_ that only Hare messages were changed, maybe I missed something?

lrettig committed:
Improvements and bugfixes to logging and context

From testnet127

lrettig committed:
Gofmt

lrettig committed:
Apply suggestions from code review

Co-authored-by: Noam Nelke <noam@spacemesh.io>

lrettig committed:
Fixes per PR review comments

Remove unused block_listener code

lrettig committed:
Small fixes

Harmonize message sender log type

lrettig committed:
Minor cleanup of syncer logging

lrettig committed:
Use contextual logger in initServices

lrettig committed:
Replace deprecated google context lib

Use stdlib context

lrettig commented:
> My main issue with this change is that it changes the Hare gossip messages on the wire for the purpose of nicer logging. I _think_ that only Hare messages were changed, maybe I missed something?

See inline replies to questions - I don't think the format of any messages has changed on the wire. These are all local only. I think all the concerns raised here have been addressed.

lrettig committed:
Pass log object around node/app object

lrettig committed:
Tidy

lrettig committed:
Merge branch 'develop' into contextual-logging

And fix merge conflicts by adding ctx to new code

lrettig committed:
Merge branch 'develop' into contextual-logging

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig committed:
Increase p2p msg propagation time

lrettig commented:
bors merge

lrettig committed:
Revert "Increase p2p msg propagation time"

This reverts commit 3702604f342a281779474fdab6ad1510efac6143.

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig committed:
Minor tweak

lrettig committed:
Fix typo

lrettig commented:
I can't understand for the life of me why the p2p systest is failing here, but it always fails with this error:

```
E       AssertionError: msg_testing: Total gossip messages in ES is not as expected
E         expected 4440, got 4300
E       assert 4440 == 4300
E         -4440
E         +4300
```

(The number of messages received varies from run to run but it's always < 4400.)

I don't see any errors or warnings in the logs. I double-checked all of the p2p-related changes here, and there are no changes to logic that I can find that would cause this sort of error - there are hardly any logic changes at all in this PR, it's almost entirely cosmetic. I tried increasing the timeout, and that also didn't help. I'm stumped, and I'm blocked on doing further investigation by a kibana issue.

lrettig committed:
Minor cleanup of logs and comments

lrettig commented:
bors merge

lrettig committed:
Minor cleanup of logic, context, and logs

lrettig committed:
Revert logic change to fetch queue

lrettig committed:
Revert "Revert logic change to fetch queue"

This reverts commit d5cc423bae46dc08ac8dec101483824a8ed132aa.

lrettig commented:
Walked through all of the code changes today. Confirmed that there are no logic changes, AFAICT, with one small exception: the timeout in fetch. I tried reverting this change, to no avail. Also confirmed that there's no change to any of the test code. I am totally stumped as to what's causing this failure. Either I'm missing something, or else it's a complex interaction, e.g., a timing issue or something. The only thing I can think of trying is a binary search to isolate the breaking change.

lrettig committed:
Variable naming

lrettig committed:
Merge branch 'develop' into contextual-logging

And resolve merge conflicts

lrettig commented:
bors merge

lrettig commented:
bors merge-

lrettig committed:
Resolve more merge conflicts

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig committed:
Merge branch 'develop' into contextual-logging

Resolve several merge conflicts

lrettig committed:
Minor test cleanup

lrettig committed:
Fix a nasty merge conflict

lrettig committed:
Fix typo

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig committed:
Merge branch 'develop' into contextual-logging

Resolves merge conflicts

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig committed:
Improve net stack logging

lrettig commented:
bors merge

lrettig committed:
Tweak p2p test to pass

For some reason, it fails with 40 pods, but works with fewer.
Slightly improve its output for debugging.

lrettig committed:
Update p2p test params again

Increase the node count back to the default value, decrease the number
of messages. This seems to pass more reliably.

lrettig commented:
bors merge

lrettig commented:
bors merge-

lrettig commented:
bors merge

lrettig commented:
bors merge
